data-pipeline/src/trace_exporter/agent_response.rs

```diff
@@ -1,122 +1,15 @@
 // Copyright 2024-Present Datadog, Inc. https://www.datadoghq.com/
 // SPDX-License-Identifier: Apache-2.0
 
-use serde::Deserialize;
-use serde_json::{Map, Value};
-use std::io::Error as IoError;
-use std::io::ErrorKind as IoErrorKind;
-
-use crate::trace_exporter::error::TraceExporterError;
-use std::{f64, str::FromStr};
-
-/// `Rates` contains all service's sampling rates returned by the agent.
-#[derive(Debug, Deserialize)]
-pub struct Rates {
-    rate_by_service: Map<String, Value>,
-}
-
-impl Rates {
-    /// Get the sampling rate for a service and evironment pair.
-    pub fn get(&self, service: &str, env: &str) -> Result<f64, IoError> {
-        for (id, value) in &self.rate_by_service {
-            let mut it = id
-                .split(',')
-                .filter_map(|pair| pair.split_once(':'))
-                .map(|(_, value)| value);
-
-            let srv_pair = (it.next().unwrap_or(""), it.next().unwrap_or(""));
-            if srv_pair == (service, env) {
-                return value
-                    .as_f64()
-                    .ok_or(IoError::from(IoErrorKind::InvalidData));
-            }
-        }
-        // Return default
-        if let Some(default) = self.rate_by_service.get("service:,env:") {
-            default
-                .as_f64()
-                .ok_or(IoError::from(IoErrorKind::InvalidData))
-        } else {
-            Err(IoError::from(IoErrorKind::NotFound))
-        }
-    }
-}
-
-impl FromStr for Rates {
-    type Err = TraceExporterError;
-    fn from_str(s: &str) -> Result<Self, Self::Err> {
-        let obj: Rates = serde_json::from_str(s)?;
-        Ok(obj)
-    }
-}
-
 /// `AgentResponse` structure holds agent response information upon successful request.
 #[derive(Debug, PartialEq)]
-#[repr(C)]
 pub struct AgentResponse {
```

Comment:
I would use `AgentResponse` directly in the FFI layer as an opaque data type in the same way we manage the TraceExporter handle. The `AgentResponse` can hold internally a `CString` property which can be returned to the C side through an getter and, in case we would need to parse the its contents in the future in the Rust side we could get an slice without an additional cost.
IMO this way is cleaner because we don't expose all the internals to the user and will allow us to change it in future iterations in case we need other solutions in which we do the parsing. Also, in case of errors, we don't force to the user to allocate a new structure which won't be used.
WDYT?